core/src/main/scala/org/apache/spark/sql/delta/deletionvectors/RoaringBitmapArray.scala

```diff
@@ -0,0 +1,481 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.deletionvectors
+
+import java.io.IOException
+import java.nio.{ByteBuffer, ByteOrder}
+
+import scala.collection.immutable.NumericRange
+
+import com.google.common.primitives.{Ints, UnsignedInts}
+import org.roaringbitmap.{RelativeRangeConsumer, RoaringBitmap}
+
+/**
+ * A 64-bit extension of [[RoaringBitmap]] that is optimized for cases that usually fit within
+ * a 32-bit bitmap, but may run over by a few bits on occasion.
+ *
+ * This focus makes it different from [[org.roaringbitmap.longlong.Roaring64NavigableMap]] and
+ * [[org.roaringbitmap.longlong.Roaring64Bitmap]] which focus on sparse bitmaps over the whole
+ * 64-bit range.
+ *
+ * Structurally, this implementation simply uses the most-significant 4 bytes to index into
+ * an array of 32-bit [[RoaringBitmap]] instances.
+ * The array is grown as necessary to accommodate the largest value in the bitmap.
+ *
+ * *Note:* As opposed to the other two 64-bit bitmap implementations mentioned above,
+ *         this implementation cannot accommodate `Long` values where the most significant
+ *         bit is non-zero (i.e., negative `Long` values).
+ *         It cannot even accommodate values where the 4 high-order bytes are `Int.MaxValue`,
+ *         because then the length of the `bitmaps` array would be a negative number
+ *         (`Int.MaxValue + 1`).
+ */
+final class RoaringBitmapArray extends Equals {
+  import RoaringBitmapArray._
+
+  private var bitmaps: Array[RoaringBitmap] = Array.empty
+
+  /**
+   * Add the value to the container (set the value to `true`),
+   * whether it already appears or not.
+   */
+  def add(value: Long): Unit = {
+    require(value >= 0 && value <= MAX_REPRESENTABLE_VALUE)
+    val (high, low) = decomposeHighLowBytes(value)
+    if (high >= bitmaps.length) {
+      extendBitmaps(newLength = high + 1)
+    }
+    val highBitmap = bitmaps(high)
+    highBitmap.add(low)
+  }
+
+  /** Add all `values` to the container. */
+  def addAll(values: Long*): Unit = values.foreach(add)
+
+  /** Add all values in `range` to the container. */
+  def addRange(range: Range): Unit = {
+    require(0 <= range.start && range.start <= range.end)
+    if (range.isEmpty) return // Nothing to do.
+    if (range.step != 1) {
+      // Can't optimize in this case.
+      range.foreach(i => add(UnsignedInts.toLong(i)))
+      return
+    }
+    // This is an Int range, so it must fit completely into the first bitmap.
+    if (bitmaps.isEmpty) {
+      extendBitmaps(newLength = 1)
+    }
+    val end = if (range.isInclusive) range.end + 1 else range.end
+    bitmaps.head.add(range.start, end)
+  }
+
+  /** Add all values in `range` to the container. */
+  def addRange(range: NumericRange[Long]): Unit = {
```

Comment:
This is a lot of methods... do you think you will need all of these methods for the full DV implementation?

Reply:
These are going to be mostly used in tests when you can generate a custom DV for a given set of indices. Let me mark it as for tests only.